cunumeric/array.py

```diff
@@ -3483,12 +3485,28 @@ def _perform_unary_reduction(
         src,
         axis=None,
         dtype=None,
+        res_dtype=None,
         out=None,
         keepdims=False,
         args=None,
         initial=None,
         where=True,
     ):
+        # When 'res_dtype' is not None, the input and output of the reduction
+        # have different types. Such reduction operators don't take a dtype of
+        # the accumulator
+        if res_dtype is not None:
+            assert dtype is None
+            dtype = src.dtype
+        else:
+            # If 'dtype' exists, that determines both the accumulation dtype
+            # and the output dtype
+            if dtype is not None:
+                res_dtype = dtype
+            else:
+                dtype = src.dtype
+                res_dtype = src.dtype
```

Comment:
According to the NumPy documentation on `reduce`:

```
dtypedata-type code, optional
The type used to represent the intermediate results. Defaults to the data-type of the output array if this is provided, or the data-type of the input array if no output array is provided.
```

this should do one more check:

```
elif out is not None:
  dtype = out.dtype
  res_dtype = out.dtype
else:
  dtype = src.dtype
  res_dtype = src.dtype
```

Reply:
c5d7f85afd505eeba09c2d65e1f7b05d16f104ef